src/mozilla_taskgraph/transforms/scriptworker/signing.py

```diff
@@ -0,0 +1,158 @@
+)
+from voluptuous import ALLOW_EXTRA, Any, Optional, Required
+
+SIGNING_FORMATS = ["autograph_gpg"]
+SIGNING_TYPES = ["dep", "release"]
+DETACHED_SIGNATURE_EXTENSION = ".asc"
+
+signing_schema = Schema(
+    {
+        Required("attributes"): {
+            Optional("artifacts"): dict,
+            Required("build-type"): str,
+        },
+        Required("signing"): optionally_keyed_by(
+            "build-type",
+            "level",
+            {
+                Required("format"): optionally_keyed_by(
+                    "build-type", "level", Any(*SIGNING_FORMATS)
+                ),
+                Optional("type"): optionally_keyed_by(
+                    "build-type", "level", Any(*SIGNING_TYPES)
+                ),
+                Optional("ignore-artifacts"): list,
+            },
+        ),
+        Required("worker"): {
+            Required("upstream-artifacts"): [
+                {
+                    # Paths to the artifacts to sign
+                    Required("paths"): [str],
+                }
+            ],
+        },
+    },
+    extra=ALLOW_EXTRA,
+)
+
+transforms = TransformSequence()
+transforms.add_validate(signing_schema)
+
+
+@transforms.add
+def resolve_signing_keys(config, tasks):
+    for task in tasks:
+        for key in (
+            "signing",
+            "signing.format",
+            "signing.type",
+        ):
+            resolve_keyed_by(
+                task,
+                key,
+                item_name=task["name"],
+                **{
+                    "build-type": task["attributes"]["build-type"],
+                    "level": config.params["level"],
+                },
+            )
+        yield task
+
+
+@transforms.add
+def set_signing_attributes(_, tasks):
+    for task in tasks:
+        task["attributes"]["signed"] = True
+        yield task
+
+
+@transforms.add
+def set_signing_format(_, tasks):
+    for task in tasks:
+        for upstream_artifact in task["worker"]["upstream-artifacts"]:
+            upstream_artifact["formats"] = [task["signing"]["format"]]
+        yield task
+
+
+@transforms.add
+def set_signing_and_worker_type(config, tasks):
+    for task in tasks:
+        signing_type = task["signing"].get("type")
+        if not signing_type:
+            signing_type = "release" if config.params["level"] == "3" else "dep"
+
+        task.setdefault("worker", {})["signing-type"] = f"{signing_type}-signing"
+
+        if "worker-type" not in task:
+            worker_type = "signing"
+            build_type = task["attributes"]["build-type"]
+
+            if signing_type == "dep":
+                worker_type = f"dep-{worker_type}"
+            if build_type == "macos":
+                worker_type = f"{build_type}-{worker_type}"
+            task["worker-type"] = worker_type
+
+        yield task
+
+
+@transforms.add
+def filter_out_ignored_artifacts(_, tasks):
+    for task in tasks:
+        ignore = task["signing"].get("ignore-artifacts")
```

Comment:
Is there a reason we have to ignore specific things (as opposed to having the upstream tasks explicitly define what needs signing)? It's easier for debugging and other inspection to have the things we're actually going to be signing as the explicit part IMO. Not a blocker (and I'm guessing there's some reason it has to be done this way...), just curious.

Reply:
I think the most important part is that it should be an allowlist instead of a denylist. With that in mind, I think an explicit list of names, a regex of things to sign, or even a "sign all artifacts" mode are all fine.

So - what you're proposing sounds great!